### PR TITLE
[Patch] Add loss for ORT inference

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

*  Wrap `OnnxConfig` by `wrap_onnx_config_for_loss` to obtain the loss while using `ORTTrainer` under the mode `inference_with_ort=True`.
* Enable `deepspeed` for ONNX Runtime training. (Tested with ZeRO stage 2,  full availability under progress)
* Clean up unused dependencies in `ORTTrainer`.
* Update CI of onnxruntime training.
* Update associated tests.

### CODE REVIEW

**Comment by lewtun** [comment]
This line is why the doc build is currently failing: https://github.com/huggingface/optimum/pull/152/files#diff-3e928ce0b52f617b86cd0df9399c6bbb5804d6269c6a04b486613eb929449256R25

Until now we never actually imported `OnnxConfigWithLoss` etc and doing so triggers an import error because `transformers` is pinned to <4.17 in the doc build (to have both `intel` and `onnxruntime` packages in the same env):

```
from optimum.onnxruntime.trainer import ORTTrainer

ImportError: cannot import name 'TensorType' from 'transformers.utils' (/home/lewis/miniconda3/envs/optimum/lib/python3.8/site-packages/transformers/utils/__init__.py)
```

The solution is to refactor the doc build so that we build `intel` and `onnxruntime` in separate envs. Doing so will also allow us to build the Graphcore & other hardware partner docs as well. I don't have bandwidth for this right now, but happy to review a PR if someone else has time to tackle this!

**Comment by lewtun** [comment]
FYI @JingyaHuang if you want to test that the docs build locally you can run:

```
pip install ".[dev,intel.onnxruntime]'
pip install git+https://github.com/huggingface/doc-builder.git
doc-builder build optimum docs/source --build_dir test-docs --version v1.0.0 --clean
```

You'll need a Linux machine for this since one cannot install `intel` on macOS

File: optimum/onnxruntime/trainer.py [hunk 1]
```diff
@@ -1162,35 +1138,21 @@ def prediction_step_ort(
...
+                raise NotImplementedError(
+                    "Sagemaker's distrubuted data parallel features are not supported by `ORTTrainer` yet. Stay tuned!"
+                )
             else:
                 if has_labels:
                     with self.autocast_smart_context_manager():
                         loss, outputs = self.compute_loss_ort(
                             model, inputs, input_names, output_names, return_outputs=True
                         )
+                        loss = torch.tensor(loss).mean().detach()
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:1150]: Why do you use `detach()` ?

File: optimum/onnxruntime/trainer.py [hunk 2]
```diff
@@ -784,24 +749,33 @@ def evaluation_loop_ort(
...
             )
-            logger.info("------------------Exporting ORT model to ONNX IR----------------------")
-            self._export(onnx_model_path)
+            logger.info("[INFO] Exporting the model to ONNX...")
+            if has_labels:
+                self._export(onnx_model_path)
+                self.exported_with_loss = True
+            else:
+                self._export(onnx_model_path, with_loss=False)
+                self.exported_with_loss = False
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:774]:
```suggestion
            self._export(onnx_model_path, with_loss=has_labels)
            self.exported_with_loss = has_labels
```

File: optimum/onnxruntime/trainer.py [hunk 3]
```diff
@@ -432,11 +417,7 @@ def train(
             elif isinstance(train_dataloader.dataset, IterableDatasetShard):
                 train_dataloader.dataset.set_epoch(epoch)
 
-            if is_torch_tpu_available():
-                parallel_loader = pl.ParallelLoader(train_dataloader, [args.device]).per_device_loader(args.device)
-                epoch_iterator = parallel_loader
-            else:
-                epoch_iterator = train_dataloader
+            epoch_iterator = train_dataloader
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:420]: Is there a reason we introduce the variable `epoch_iterator` ? (can we just use `train_dataloader` ?)

File: optimum/onnxruntime/trainer.py [hunk 4]
```diff
@@ -85,27 +86,15 @@
 import onnx
 import onnxruntime
 
-from .utils import fix_atenops_to_gather  # , _is_gpu_available # wait for the merge of ORTMoel
+from .utils import (  # , _is_gpu_available # wait for the merge of ORTMoel
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:89]: should we remove this comment ?

**Comment by echarlaix** [comment]
This PR looks great !

File: optimum/onnxruntime/trainer.py [hunk 5]
```diff
@@ -1231,43 +1193,56 @@ def compute_loss_ort(self, model, inputs, input_names, output_names, return_outp
...
         input_feed = dict(map(lambda input_name: (input_name, inputs[input_name].cpu().numpy()), input_names))
         outputs = self.infer_sess.run(output_names, input_feed)
-        loss = None
         # Save past state if it exists
         # TODO: this needs to be fixed and made cleaner later.
         if self.args.past_index >= 0:
             self._past = outputs[self.args.past_index]
 
+        if labels is not None:
+            loss = self.label_smoother(outputs, labels)
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:1207]: A fix on the compatibility of the label smoother has just been submitted.

File: optimum/onnxruntime/trainer.py (hunk 1 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py]: There is no need to detach here, will remove it.

**Pull request merged**